### PR TITLE
Yoshi serverless tests

### PR DESCRIPTION
**Pull request by yanivefraim**
Adding tests to Yoshi Serverless (after successfully finishing the POC)

### CODE REVIEW

File: test/scripts.ts [hunk 1]
```diff
@@ -239,11 +249,15 @@ export default class Scripts {
           );
         }),
         Promise.all([
-          this.projectType !== 'flow-library'
+          this.projectType !== 'flow-library' &&
+          this.projectType !== 'yoshi-serverless-typescript'
             ? waitForPort(this.serverProcessPort, { timeout: 60 * 1000 })
             : Promise.resolve(),
+          this.projectType === 'yoshi-serverless-typescript'
+            ? waitForPort(7777, { timeout: 60 * 1000 })
```
Comment (yanivefraim) [inline test/scripts.ts:257]: should read the port from some place?

File: test/scripts.ts [hunk 2]
```diff
@@ -224,13 +234,13 @@ export default class Scripts {
...
     try {
       await Promise.race([
-        waitForStdout(startProcess, 'Compiled with warnings', {
-          throttle: true,
-        }).then(() => {
-          throw new Error(
-            `Yoshi start was compiled with warnings \n \n ${startProcessOutput}`,
-          );
-        }),
+        // waitForStdout(startProcess, 'Compiled with warnings', {
```
Comment (yanivefraim) [inline test/scripts.ts:237]: fix warnings and put back

File: test/scripts.ts [hunk 3]
```diff
@@ -239,11 +249,15 @@ export default class Scripts {
...
+          this.projectType !== 'flow-library' &&
+          this.projectType !== 'yoshi-serverless-typescript'
             ? waitForPort(this.serverProcessPort, { timeout: 60 * 1000 })
             : Promise.resolve(),
+          this.projectType === 'yoshi-serverless-typescript'
+            ? waitForPort(7777, { timeout: 60 * 1000 })
+            : Promise.resolve(),
           waitForPort(this.staticsServerPort, { timeout: 60 * 1000 }),
-          waitForStdout(startProcess, 'Compiled successfully!'),
+          // waitForStdout(startProcess, 'Compiled successfully!'),
```
Comment (yanivefraim) [inline test/scripts.ts:260]: fix warnings and put back

File: test/typescript/features/yoshi-server/api-client-serverless/api-client-serverless.test.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  projectType: 'yoshi-serverless-typescript',
+});
+
+describe.each(['dev'] as const)(
+  'yoshi-server api client to server [%s]',
+  (mode) => {
+    it('run tests', async () => {
+      await scripts[mode](async () => {
+        await page.goto(
+          `http://localhost:7777/serverless/kitchensink-0-0-0/app`,
```
Comment (yanivefraim) [inline test/typescript/features/yoshi-server/api-client-serverless/api-client-serverless.test.ts:16]: should read the url from somewhere

File: test/typescript/features/yoshi-server/api-client-serverless/src/templates/app.ejs [hunk 5]
```diff
@@ -0,0 +1,14 @@
+<!doctype html>
```
Comment (yanivefraim) [inline test/typescript/features/yoshi-server/api-client-serverless/src/templates/app.ejs:1]: we can use the fixture template here

File: test/yoshi-publish/package.json [hunk 6]
```diff
@@ -28,6 +30,7 @@
     "@types/puppeteer": "1.20.3",
     "@types/react": "16.9.38",
     "@types/react-dom": "^16.0.11",
+    "@wix/serverless-testkit": "^1.0.0",
```
Comment (yanivefraim) [inline test/yoshi-publish/package.json:33]: check if it should be here or @ root package.json (or both?)

File: packages/yoshi-serverless/src/render.ts [hunk 7]
```diff
@@ -0,0 +1,18 @@
+import path from 'path';
```
Comment (ronami) [inline packages/yoshi-serverless/src/render.ts:1]: we should probably make sure it works in production or if it needs to be copied somewhere

File: test/fixtures/yoshi-serverless-typescript/base-template/environment.js [hunk 8]
```diff
@@ -0,0 +1,11 @@
+const { app } = require('@wix/serverless-testkit');
+const {
+  getServerlessScope,
+} = require('../../../../../packages/yoshi-helpers/build/utils');
```
Comment (ronami) [inline test/fixtures/yoshi-serverless-typescript/base-template/environment.js:4]: can probably change to `require('yoshi-helpers/build/utils')`

File: test/scripts.ts [hunk 9]
```diff
@@ -54,29 +57,37 @@ export default class Scripts {
...
+    this.ignoreWarnings = ignoreWarnings;
     this.verbose = !!process.env.DEBUG;
     this.testDirectory = testDirectory;
     this.serverProcessPort = 3000;
     this.staticsServerPort = projectType === 'flow-library' ? 3300 : 3200;
     this.storybookServerPort = 9009;
     this.serverUrl = `http://localhost:${this.serverProcessPort}`;
+    this.serverlessUrl = `http://localhost:${
+      this.serverProcessPort
+    }/serverless/${getServerlessScope(testDirectory)}`;
```
Comment (ronami) [inline test/scripts.ts:90]: developers that use routes will also need to do this, we don't focus routes now but I think users (and e2e tests) should be able to access `http://localhost:3000` and access their routes.

File: test/scripts.ts [hunk 10]
```diff
@@ -129,12 +142,17 @@ export default class Scripts {
       yoshiBinToUse = yoshiFlowLibraryBin;
     }
 
+    let envVars = '';
+    if (projectType === 'yoshi-serverless-typescript') {
+      envVars = 'EXPERIMENTAL_YOSHI_SERVERLESS=true ';
+    }
```
Comment (ronami) [inline test/scripts.ts:148]: I think we should use an object, can be a bit clearer, WDYT?

File: test/scripts.ts [hunk 11]
```diff
@@ -220,30 +242,40 @@ export default class Scripts {
...
+        ...errorsArray,
         Promise.all([
           this.projectType !== 'flow-library'
             ? waitForPort(this.serverProcessPort, { timeout: 60 * 1000 })
             : Promise.resolve(),
           waitForPort(this.staticsServerPort, { timeout: 60 * 1000 }),
-          waitForStdout(startProcess, 'Compiled successfully!'),
+          !this.ignoreWarnings
+            ? waitForStdout(startProcess, 'Compiled successfully!')
+            : Promise.resolve<string>(''),
```
Comment (ronami) [inline test/scripts.ts:278]: `<string>` can probably be removed

File: test/typescript/features/yoshi-server/api-client-serverless/api-client-serverless.test.ts [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+
+jest.setTimeout(50 * 1000);
+
+const scripts = Scripts.setupProjectFromTemplate({
+  templateDir: __dirname,
+  projectType: 'yoshi-serverless-typescript',
+  ignoreWarnings: true,
+});
+
+describe.each(['dev'] as const)(
```
Comment (ronami) [inline test/typescript/features/yoshi-server/api-client-serverless/api-client-serverless.test.ts:11]: can we do 'prod' too? we can do in a future PR too

**Pull request merged**